### PR TITLE
Remove (deprecated) BTE Framework from 7.x codebase

### LINKED ISSUE #2817: DS-4491: Remove (deprecated) BTE Framework from 7.x codebase
This is a duplicate of [DS-4491](https://jira.lyrasis.org/browse/DS-4491)

https://wiki.lyrasis.org/display/DSPACE/BTE+transition%3A+4Science+implementation+proposal+and+estimation

**Estimate from release plan: TBD
Expressing interest: 4Science**

The BTE (Biblio-Transformation-Engine) Framework was officially deprecated in DSpace 6.x. This framework was used by the JSPUI for importing metadata from external sources and/or common structured files (like BibTeX, etc). This framework was built/maintained by the Hellenic National Documentation Centre (EKT)

Deprecation (and removal) is necessary for the following reasons:

The EKT developer who built this framework left their organization in/around 2015. While EKT staff said they'd try to support it for one more release (v6.x), it's long term future was uncertain.
Since 2014, no maintenance/updates have occurred on BTE: https://github.com/EKT/Biblio-Transformation-Engine Obviously, DSpace cannot use continue to use third-party libraries that are unmaintained.
The BTE codebase has become stale and uses additional libraries with known security vulnerabilities. See comment below for full details.
The recommendation of this ticket is to remove BTE in favor of the "Live Import" framework (which overlaps in functionality, but was XMLUI specific in v6). See the "Live Import" 6.x docs at:

General overview: https://wiki.lyrasis.org/pages/viewpage.action?pageId=68064788#Introduction
More details: https://wiki.lyrasis.org/display/DSDOC6x/2016+Framework+for+live+import+from+external+sources
This choice of a single importer framework also aligns with the goals laid out in our last Strategic Plan, namely to be "lean" and avoid duplicative functionality in our codebase.

Before removing BTE, we should look to replace each of its features with a Live Import based version. Full replacement need not be required for BTE removal, but ideally there is minimal (or zero) loss of functionality for old JSPUI-based users (XMLUI-based users would have no functionality loss as BTE never supported XMLUI). Additional tickets should be linked to this one describing each of the missing gaps in "Live Import" that are necessary to replace BTE.

This ticket has been tentatively scheduled for 7.0, as we should not encourage institutions upgrading to 7.0 to continue to utilize BTE. Ideally, therefore, it should be removed prior to the 7.0 final release, in order to no longer support two parallel frameworks.

### LINKED ISSUE #3012: DOI metadata extraction in Submission should be disabled and/or refactored to use external sources
If you enter a DOI in the submission form, the backend will attempt to lookup that DOI to extract metadata and add that metadata on the server-side to the active submission.  This uses BTE for the lookup, which will be removed in #2817

The idea of enhancing (or starting) a submission by looking up a DOI is a good one.  

However, this current implementation is disruptive to the submitter:
* The DOI field is midway down the page, so it's highly likely that the submitter has entered most metadata manually. 
* Because of that, once the DOI lookup occurs, any differences in metadata (between manually entered and automatically discovered via DOI) need to be addressed or merged.  This results in currently odd behaviors...e.g. if the DOI finds a different value, it adds it as a separate value to the existing field, _even if that field is only allowed one value._  For example, if an alternative Title is discovered, you could end up with two values in the Title field (even if only one is supported).  This results in an invalid submission form, until a value is removed.

More recently, we now support this same sort of "lookup" in external services via the [`/externalsources` endpoint](https://github.com/DSpace/Rest7Contract/blob/main/external-authority-sources.md) and in the UI by clicking on the "Import metadata from external sources" button _prior to starting a submission._

Therefore, I propose that we do the following:
1. Remove the existing DOI lookup feature which is triggered when entering a DOI in the submission form.  This will be necessary anyway as part of #2817, as this feature is specific to BTE at this time.
2. Add DOI lookup to the existing `/externalsources`, so that a submitter can choose to *start a new submission via a DOI* from the MyDSpace page.

This will solve the issues above by removing the DOI lookup during an active submission.  At a later time, we can revisit this & potentially add this feature back in as part of https://github.com/DSpace/dspace-angular/issues/896 (revisiting auto-save behaviors).  Essentially, dynamically inserting metadata into the submission form will require "smarter" auto-save behaviors which allow a submitter to see & resolve conflicts or differences between manually-added metadata & dynamically-added metadata.

### PR DESCRIPTION
**Pull request by abollini**
## References
* Fixes #2817
* Fixes #3012

## Description
All the references to the BTE framework has been removed from the code base. The extract metadata step has been modified to use the Live Import Framework and all the code related to enhancement of current workspace item both via identifiers than via file has been centralized in such step. This step is NOT part of the default configuration, users need to enable it explicitly.

Finally some cleanup to the submission steps hierarchy have been performed and the lifecycle of the patch/upload processing for inprogress items have been fixed so that all the steps participate in the preProcessing phase and, after than the target step has processed the request, they are involved again in the postProcessing.

## Instructions for Reviewers
Review the changes in the WorkspaceItemRestIT the test added cover the extract metadata step that previously was not tested.

To test manually the metadata-extract step you need to uncomment the step in the submission-forms.xml

adding a pmid to the dc.identifier.other should lead to the automatic enrichment of the item with information retrieved from PubMed.

Test the submission features at large to verify that the BTE removal has had no side effect (but the large number of ITs should make us safe)

## Checklist

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [N/A] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [N/A] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Review by tdonohue — CHANGES_REQUESTED** [review]
@abollini : Gave this a review today... Overall, it looks good, and I'm glad to see tests are provided for the new code (the new, disabled-by-default, extraction step).  I found a few very minor issues that need cleanup.

I also tried manually testing the metadata extraction step....while the tests show that it *should* work, it doesn't appear to work from the UI.  Also, enabling it blocks the ability to complete a submission, as the UI shows this step as being required (and empty):
extract-step

I also cannot seem to trigger the step to take action (I tried uploading a BibTex file, and also adding a PubMed ID to dc.identifier.other and neither worked)

I think we need to do one of the following with this PR:
1. Either we remove the Extraction Step code/configurations from this PR & create a follow-up PR with that work  (This allows us to move the BTE Framework removal forward and finish up the Extraction Step later)
OR
2. Merge it as-is, but add a warning that this Extraction Step does NOT currently work from the UI.  Then, we'd have to follow up later with UI fixes to allow the Extraction Step to work.

I lean slightly towards (1), as the Extraction Step is not required at this time (it's not in the 7.0 scope, instead it's related to https://github.com/DSpace/dspace-angular/issues/896 which is scheduled for 7.1), but the BTE removal *is required*.

File: dspace-api/pom.xml [hunk 1]
```diff
@@ -537,17 +537,17 @@
...
-                <exclusion>
-                    <groupId>net.bytebuddy</groupId>
-                    <artifactId>byte-buddy</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
+		<dependency>
+			<groupId>net.sf.opencsv</groupId>
+			<artifactId>opencsv</artifactId>
+			<version>2.3</version>
```
[the remark this answers is not in the document]
Comment (mwoodiupui) [reply dspace-api/pom.xml:543]: See https://mvnrepository.com/artifact/net.sf.opencsv/opencsv where it says "Note: This artifact was moved to:  com.opencsv » opencsv"  The two have different `groupId` but are the same project, ten years (and two repositories) apart.  We should use at least 5.2.

**Review by KevinVdV — CHANGES_REQUESTED** [review]
I made an initial review of the code see comment above, aside from that comment I have the following issues/questions:

* The tests are currently making calls to the pubmed service itself, that should be mocked
* I noticed that you removed a couple of classes in the "org.dspace.submit.step" package in the dspace-api, is there a reason why the other ones aren't removed ? They don't appear to contain any code is there a reason to not remove that entire package ?
* I also couldn't get the extraction of metadata to work from the UI

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/ExtractMetadataStep.java [hunk 2]
```diff
@@ -9,105 +9,207 @@
...
+        return changedMetadata;
+    }
+
     @Override
     public ErrorRest upload(Context context, SubmissionService submissionService, SubmissionStepConfig stepConfig,
             InProgressSubmission wsi, MultipartFile multipartFile)
         throws IOException {
 
         Item item = wsi.getItem();
+        File file = Utils.getFile(multipartFile, "extract-metadata-step", stepConfig.getId());
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/ExtractMetadataStep.java:165]: all the uploadable steps are invoked when a file is uploaded, the "extract-metadata-step" prefix is only used for the temporary files created for the step processing

**Comment by abollini** [comment]
@tdonohue @KevinVdV all your feedback should have been addressed. I haven't tested yet with a running Angular UI we can do that tomorrow but I would appreciate if you are able to quickly retest and let me know. As the extract step is expected to be invisible to the UI (and now I have added a test to verify - and fix the previous mistake) everything should be ok. The PR is beome a bit larger as I have add to add the pubmed import service mock, extra tests and refactor other part of the submission code removing unused class and centralizing some duplicate code among WorkspaceItemRestRepository and WorkflowItemRestRepository

File: dspace-api/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/pom.xml]: done

File: dspace-api/pom.xml [hunk 3]
```diff
@@ -537,17 +537,17 @@
...
+			<scope>compile</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.jbibtex</groupId>
+			<artifactId>jbibtex</artifactId>
+			<version>1.0.10</version>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/pom.xml]: done

File: dspace-api/src/test/data/dspaceFolder/config/item-submission.xml [hunk 4]
```diff
@@ -81,20 +82,16 @@
...
+            <type>cclicense</type> 
+        </step-definition>
 
-        <!--Step will be to Verify/Review everything -->
-        <!-- <step-definition id="verify"> <heading>submit.progressbar.verify</heading> <processing-class>org.dspace.submit.step.VerifyStep</processing-class>
-            <type>verify</type> </step-definition> -->
+        <step-definition id="extractionstep"> 
+            <heading>submit.progressbar.ExtractMetadataStep</heading>
+            <processing-class>org.dspace.app.rest.submit.step.ExtractMetadataStep</processing-class>
+            <type>cclicense</type>
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/test/data/dspaceFolder/config/item-submission.xml]: done

File: dspace/config/item-submission.xml [hunk 5]
```diff
@@ -119,14 +119,15 @@
...
-            <processing-class>org.dspace.submit.step.DetectPotentialDuplicate</processing-class> 
-            <type>duplicate</type> </step-definition> -->
-
-        <!--Step will be to Verify/Review everything -->
-        <!-- <step-definition id="verify"> <heading>submit.progressbar.verify</heading> <processing-class>org.dspace.submit.step.VerifyStep</processing-class>
-            <type>verify</type> </step-definition> -->
+        <!--Step will be to enrich the current submission querying external providers or processing the uploaded file -->
+        <!-- Uncomment this step to enrich the current submission using information extracted 
+            from uploaded files or metadata. Please check also config/spring/api/step-processing-listener.xml
+            for further configuration -->
```
Comment (tdonohue) [inline dspace/config/item-submission.xml:125]: Just a note which we may want to add to this configuration.  It looks like, currently, this step is only triggered during either:
* An **upload** (e.g. extract metadata from a BibTex file)
* A **save** of the form. (e.g. entering a PubMed ID only triggers a lookup if you click "Save" or "Save for later")

**Review by tdonohue — APPROVED** [review]
👍 @abollini : This looks good to me now.  

* The BTE functionality is removed (and I'm not noticing any obvious side-effects)
* The new automated metadata extraction step is now invisible (not shown in the Submission form) and it's _disabled by default_ as [we recommended in ticket 3012](https://github.com/DSpace/DSpace/issues/3012#issuecomment-714730029)

I've also successfully tested the metadata extraction step from the Submission UI.  However, it's worth noting that it's only triggered during a file upload or saving of the form.  So, we should document that (see my inline note below), especially since the form no longer auto-saves by default.

In any case, this looks good to me.  Thanks!

File: dspace/config/item-submission.xml (hunk 5 above)
Comment (abollini) [reply dspace/config/item-submission.xml]: done. I have also mentioned the autosave feature available on the angular side that can be used with that to make the process more user friendly

**Comment by tdonohue** [comment]
@KevinVdV : Could you give this another quick review?  It looks (to me) like your prior feedback has all been addressed.  But, I'd prefer if you took a second look before merger. Thanks!

**Review by KevinVdV — COMMENTED** [review]
My issues with the code have been fixed, so on a code basis I can approve the PR. However when testing it I ran into some issues.

I enabled the "extractionstep" in the "item-submission.xml" by:
* Commenting out the "step-definition" section
* Adding the step in the "submission-process" section

I then started a new submission & added the "32373993" value to the dc.identifier.other field & pressed the "save" button. I then got a 500 error response code from the PATCH request.

I found the following exception in the dspace logs:

`
	at org.dspace.importer.external.service.components.AbstractRemoteMetadataSource.retry(AbstractRemoteMetadataSource.java:179) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.importer.external.pubmed.service.PubmedImportMetadataSourceServiceImpl.getRecord(PubmedImportMetadataSourceServiceImpl.java:132) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.external.provider.impl.LiveImportDataProvider.getExternalDataObject(LiveImportDataProvider.java:85) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.submit.listener.SimpleMetadataListener.getExternalDataObject(SimpleMetadataListener.java:62) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.app.rest.submit.step.ExtractMetadataStep.doPostProcessing(ExtractMetadataStep.java:97) ~[classes/:?]
	... 131 more
Caused by: org.dspace.importer.external.exception.MetadataSourceException: At retry of operation 47270d87-e605-49db-8f92-6d706224eef3 0
	at org.dspace.importer.external.service.components.AbstractRemoteMetadataSource.throwSourceException(AbstractRemoteMetadataSource.java:248) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.importer.external.service.components.AbstractRemoteMetadataSource.retry(AbstractRemoteMetadataSource.java:168) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.importer.external.pubmed.service.PubmedImportMetadataSourceServiceImpl.getRecord(PubmedImportMetadataSourceServiceImpl.java:132) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.external.provider.impl.LiveImportDataProvider.getExternalDataObject(LiveImportDataProvider.java:85) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.submit.listener.SimpleMetadataListener.getExternalDataObject(SimpleMetadataListener.java:62) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.app.rest.submit.step.ExtractMetadataStep.doPostProcessing(ExtractMetadataStep.java:97) ~[classes/:?]
	... 131 more
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: com.sun.ws.rs.ext.RuntimeDelegateImpl
`

It appears to be an issue with a library somewhere ?

Tomcat I used: 9.0.29
Java version: 11.0.2

**Comment by mwoodiupui** [comment]
That class should have been found in jersey-common.

**Comment by tdonohue** [comment]
@abollini : There's a new small merge conflict here...I think it was caused by a minor change to the dspace-api/pom.xml in #3171 (which was just merged).

In the meantime, I'll see if I can reproduce the dependency issue that @KevinVdV [noted above](https://github.com/DSpace/DSpace/pull/3189#pullrequestreview-628720183).

**Comment by mwoodiupui** [comment]
Since it appears to be trying date formats one after another until one succeeds, perhaps those INFOs should be DEBUGs.

**Comment by abollini** [comment]
I have resolved the merge conflict and take the chance to improve the logging as by @mwoodiupui suggestion in this commit https://github.com/DSpace/DSpace/pull/3189/commits/683cba643efa2acca27569bb988878fb72b66817
(if all the parser fail a single log info is now reported)

**Comment by abollini** [comment]
I have also performed a new manual test, similar to what @tdonohue performed and I confirm that everything works as expected. I'm not able to reproduce the issue reported by @KevinVdV

**Review by KevinVdV — APPROVED** [review]
I tried to test the extraction step on another machine & it worked there. Most be an issue with my local setup. Not going to hold this PR on that. Approving.

**Comment by tdonohue** [comment]
Thanks @abollini !  Merging this as it's now at +2.
